src/com/sun/jna/Memory.java

```diff
@@ -66,13 +125,62 @@ public static void purge() {
 
     /** Dispose of all allocated memory. */
     public static void disposeAll() {
-        // use a copy since dispose() modifies the map
-        Collection<Memory> refs = new LinkedList<Memory>(allocatedMemory.keySet());
-        for (Memory r : refs) {
-            r.dispose();
+        synchronized (LinkedReference.class) {
+            LinkedReference entry;
+
+            while ((entry = HEAD) != null) {
+                Memory memory = HEAD.get();
+
+                if (memory != null) {
+                    // dispose does the unlink call internal
+                    memory.dispose();
```

Comment:
Iterating the list just to call `dispose()` does a lot of re-linking only to delete those links in the next iteration.  I realize we probably don't care as much about the performance of this method, but could this loop not just iterate the list doing the `free()` portion of the existing `dispose()` method, but not waste any effort tracking the links along the way?